File: scrumly-ui/src/app/member-dashboard/modules/organizations/services/team.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable } from 'rxjs';
import {
  CreateTeamRQ,
  OrganizationTeamGroupDto,
  TeamDto,
  TeamMemberDto,
  UpdateTeamRQ
} from '../model/organization.model';
import { environment } from '../../../../../enviroments/enviroment';
import { PageDto, SearchQuery } from '../../../../ui-components/models/search-filter.model';

@Injectable({
  providedIn: 'root'
})
export class TeamService {
  constructor(private httpClient: HttpClient) {
  }

  // HTTP METHODS

  public createTeam(createTeamRQ: CreateTeamRQ): Observable<TeamDto> {
    const url = `${ environment.api_url }/users/team`;
    return this.httpClient.post<TeamDto>(url, createTeamRQ);
  }

  public updateTeam(teamId: string, updateTeamRQ: UpdateTeamRQ): Observable<TeamDto> {
    const url = `${ environment.api_url }/users/team/${ teamId }`;
    return this.httpClient.put<TeamDto>(url, updateTeamRQ);
  }

  public findTeams(searchQuery: SearchQuery): Observable<PageDto<TeamDto>> {
    const url = `${ environment.api_url }/users/team/all`;
    return this.httpClient.post<PageDto<TeamDto>>(url, searchQuery);
  }

  public findMyTeams(searchQuery: SearchQuery): Observable<PageDto<TeamDto>> {
    const url = `${ environment.api_url }/users/team/me`;
    return this.httpClient.post<PageDto<TeamDto>>(url, searchQuery);
  }

  public findOrganizationTeamGroup(): Observable<OrganizationTeamGroupDto[]> {
    const url = `${ environment.api_url }/users/team/groups/me`;
    return this.httpClient.get<OrganizationTeamGroupDto[]>(url);
  }

  public findTeamsByOrganizationId(orgId: string, searchQuery: SearchQuery): Observable<PageDto<TeamDto>> {
    const url = `${ environment.api_url }/users/team/organization/${ orgId }`;
    return this.httpClient.post<PageDto<TeamDto>>(url, searchQuery);
  }

  public findTeamById(teamId: string): Observable<TeamDto> {
    const url = `${ environment.api_url }/users/team/${ teamId }`;
    return this.httpClient.get<TeamDto>(url);
  }

  public deleteTeamById(teamId: string): Observable<void> {
    const url = `${ environment.api_url }/users/team/${ teamId }`;
    return this.httpClient.delete<void>(url);
  }

  public inviteTeamMembers(teamId: string, inviteMembers: string[]): Observable<TeamDto> {
    const url = `${ environment.api_url }/users/team/${ teamId }/invite`;
    return this.httpClient.post<TeamDto>(url, inviteMembers);
  }

  public removeTeamMembers(teamId: string, members: string[]): Observable<TeamDto> {
    const url = `${ environment.api_url }/users/team/${ teamId }/remove`;
    return this.httpClient.post<TeamDto>(url, members);
  }

  public checkIfExistsInTeam(teamId: string, userIds: string[]): Observable<boolean> {
    const url = `${ environment.api_url }/users/team/${ teamId }/check`;
    return this.httpClient.post<boolean>(url, userIds);
  }

}

File: scrumly-backend/scrumly-api/user-service/src/main/java/com/scrumly/userservice/userservice/services/TeamService.java
package com.scrumly.userservice.userservice.services;

import com.scrumly.specification.PageDto;
import com.scrumly.specification.SearchQuery;
import com.scrumly.userservice.userservice.dto.requests.CreateTeamRQ;
import com.scrumly.userservice.userservice.dto.requests.UpdateTeamRQ;
import com.scrumly.userservice.userservice.dto.service.organization.OrganizationTeamGroupDto;
import com.scrumly.userservice.userservice.dto.service.team.TeamDto;
import com.scrumly.userservice.userservice.dto.service.team.TeamMetadataDto;
import com.scrumly.userservice.userservice.dto.service.user.UserProfileDto;

import java.util.List;

public interface TeamService {
    TeamDto createTeam(CreateTeamRQ createTeamRQ);

    TeamDto updateTeam(String teamId, UpdateTeamRQ updateTeamRQ);

    PageDto<TeamDto> findTeams(SearchQuery searchQuery);

    PageDto<TeamDto> findMyTeams(SearchQuery searchQuery);

    PageDto<TeamDto> findTeamsByOrganizationId(SearchQuery searchQuery, String orgId);

    TeamDto findTeamById(String teamId);

    void deleteTeamById(String teamId);

    TeamDto inviteTeamMembers(String teamId, List<String> inviteMembers);

    TeamDto removeTeamMembers(String teamId, List<String> members);

    boolean checkIfExistsInTeam(String teamId, List<String> userIds);

    List<OrganizationTeamGroupDto> findOrganizationTeamGroup(String userId);

    TeamMetadataDto findTeamMetadata(String teamId);

    List<UserProfileDto> findTeamUsers(String teamId);
}
